### PR TITLE
Workaround memset alignment sensitivity

### LINKED ISSUE #24300: Array.Clear performance is sensitive to alignment
Repro:

```
using System;
using System.Runtime.InteropServices;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromTypes(new[] { typeof(Program) }).Run(args);

    static unsafe byte[] AllocateArray(int size, int unalignment)
    {
        Random r = new Random();

        for (;;)
        {
             byte[] a = new byte[size];
             GCHandle handle = GCHandle.Alloc(a, GCHandleType.Pinned);
             fixed (void *p = a) {
                 if (((int)p & 0x1F) == unalignment) return a;
             }
             handle.Free();
             GC.KeepAlive(new byte[r.Next(0x1F)]);
        }
    }

    private byte[] _arr8192 = AllocateArray(8192, 0);
    private byte[] _arr8192Unaligned = AllocateArray(8192, 8);
    [Benchmark] public void Clear8192() => Array.Clear(_arr8192, 0, _arr8192.Length);
    [Benchmark] public void Clear8192Unaligned() => Array.Clear(_arr8192Unaligned, 0, _arr8192Unaligned.Length);
}
```

Result on Windows x64:

```
BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Xeon CPU E5-2630 v3 2.40GHz, 1 CPU, 16 logical and 8 physical cores
.NET Core SDK=3.0.100-preview5-011472
  [Host]     : .NET Core 3.0.0-preview5-27619-01 (CoreCLR 4.6.27618.74, CoreFX 4.700.19.21810), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview5-27619-01 (CoreCLR 4.6.27618.74, CoreFX 4.700.19.21810), 64bit RyuJIT


|             Method |     Mean |     Error |    StdDev |
|------------------- |---------:|---------:|---------:|
|          Clear8192 | 106.8 ns | 3.309 ns | 4.850 ns |
| Clear8192Unaligned | 190.4 ns | 1.450 ns | 1.211 ns |
```

Found by @stephentoub

### PR DESCRIPTION
**Pull request by jkotas**
memset is up to 2x slower on misaligned block on some types of hardware. The problem is uneven performance of "rep stosb"
used to implement the memset in some cases. The exact matrix on when it is slower and by how much is very complex.

This change workarounds the issue by aligning the memory block before it is passed to memset and filling in the potential misaligned
part manually. This workaround will regress performance by a few percent (<10%) in some cases, but we will gain up to 2x improvement
in other cases.

Fixes #24300

### CODE REVIEW

**Comment by jkotas** [comment]
cc @stephentoub  @EgorBo

**Comment by EgorBo** [comment]
I wonder if it makes sense for big arrays to implement `memset` in the managed code via AVX if available instead of `rep stosb` (e.g. this is what clang generate for a plain C loop: https://godbolt.org/z/Tl-YFi).

E.g.:

```
    [Benchmark]
    public unsafe void Clear8192AvxUnrolled()
    {
        var zerov = Vector256<byte>.Zero;
        var c = _arr8192;
        fixed (byte* p = c)
        {
            for (int i = 0; i < c.Length; i += 32 * 4)
            {
                Avx.Store(p + i, zerov);
                Avx.Store(p + i + 32, zerov);
                Avx.Store(p + i + 64, zerov);
                Avx.Store(p + i + 96, zerov);
            }
        }
    }

    [Benchmark]
    public unsafe void Clear8192AvxUnrolledUnaligned()
    {
        var zerov = Vector256<byte>.Zero;
        var c = _arr8192;
        fixed (byte* p = c)
        {
            for (int i = 0; i < c.Length; i += 32 * 4)
            {
                Avx.Store(p + i, zerov);
                Avx.Store(p + i + 32, zerov);
                Avx.Store(p + i + 64, zerov);
                Avx.Store(p + i + 96, zerov);
            }
        }
    }
```

```
|                        Method |      Mean |     Error |    StdDev |
|------------------------------ |----------:|----------:|----------:|
|                     Clear8192 |  71.18 ns | 0.5130 ns | 0.4798 ns |
|            Clear8192Unaligned | 101.68 ns | 0.0083 ns | 0.0077 ns |
|          Clear8192AvxUnrolled |  54.55 ns | 0.0267 ns | 0.0250 ns |
| Clear8192AvxUnrolledUnaligned |  55.02 ns | 0.0139 ns | 0.0130 ns |
```

File: src/System.Private.CoreLib/shared/System/SpanHelpers.cs [hunk 1]
```diff
@@ -24,7 +24,7 @@ public static unsafe void ClearWithoutReferences(ref byte b, nuint byteLength)
                 return;
 
 #if CORECLR && (AMD64 || ARM64)
-            if (byteLength > 4096)
+            if (byteLength > 768)
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/SpanHelpers.cs]: Nit: could you add a comment as to how this number was deduced?

**Comment by benaadams** [comment]
> `Avx.Store` 

Aside; as not what this PR is doing, but `Avx.StoreAlignedNonTemporal` would be better, as clear after use is more common than clear before use?

**Comment by jkotas** [comment]
> I wonder if it makes sense for big arrays to implement memset in the managed code via AVX if available instead of rep stosb

It is a rocket science to do this well. These things need to be tuned against real workloads, microbenchmarks are not enough.

File: src/System.Private.CoreLib/shared/System/SpanHelpers.cs (hunk 1 above)
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/SpanHelpers.cs]: ran the benchmark for different array sizes:
```
|         Method | arraySize |     Mean |
|--------------- |---------- |---------:|
|          Clear |       500 | 13.00 ns |
| ClearUnaligned |       500 | 11.32 ns |
|          Clear |       600 | 16.99 ns |
| ClearUnaligned |       600 | 16.94 ns |
|          Clear |       700 | 17.21 ns |
| ClearUnaligned |       700 | 23.03 ns |
|          Clear |       800 | 17.60 ns |
| ClearUnaligned |       800 | 18.15 ns |
|          Clear |       900 | 18.02 ns |
| ClearUnaligned |       900 | 19.22 ns |
```
but these results are for the old code `if (byteLength > 4096)` so all of them use `Unsafe.InitBlockUnaligned` and the problem appears after count > 600 on my machine (i8700K Windows10 x64)

**Comment by EgorBo** [comment]
@benaadams will try but that was just a proof of concept :)

File: src/System.Private.CoreLib/shared/System/SpanHelpers.cs [hunk 2]
```diff
@@ -24,7 +24,9 @@ public static unsafe void ClearWithoutReferences(ref byte b, nuint byteLength)
                 return;
 
 #if CORECLR && (AMD64 || ARM64)
-            if (byteLength > 4096)
+            // The exact matrix on when RhZeroMemory is faster than InitBlockUnaligned is very complex. The factors to consider include
+            // type of hardware and memory aligment. This treashold was choosen as good ballance accross different configurations.
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/SpanHelpers.cs]:
```suggestion
            // type of hardware and memory aligment. This threshold was chosen as a good balance accross different configurations.
```

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/SpanHelpers.cs]: Thanks for adding the comment.

**Comment by EgorBo** [comment]
@benaadams just tried - `Avx.StoreAlignedNonTemporal` is 5x times slower for the benchmark above (aligned data, 8196 bytes). The only difference in the Asm output is `vmovntdq` instead of `vmovdqa` (or `vmovdqu`).

**Comment by jkotas** [comment]
@dotnet-bot test this please

**Comment by EgorBo** [comment]
AVX with force align:
```csharp
    public static void MemsetLargeArrayAvx(byte* dst, int length)
    {
        // this code is supposed to be called only for large arrays
        // so we don't have to care about small arrays.
        *((nuint*)dst + 0) = 0;
        *((nuint*)dst + 1) = 0;
        *((nuint*)dst + 2) = 0;
        *((nuint*)dst + 3) = 0;
        byte* dsta = (byte*)(((nuint) (dst + 1) + 31) & ~(nuint) 31);

        var zero = Vector256<byte>.Zero;
        int i = 0;
        for (; i < length - 128; i += 128)
        {
            Avx.StoreAligned(dsta + i, zero);
            Avx.StoreAligned(dsta + i + 32, zero);
            Avx.StoreAligned(dsta + i + 64, zero);
            Avx.StoreAligned(dsta + i + 96, zero);
        }

        var endElements = (uint) (length - i);
        if (endElements > 0)
            Unsafe.InitBlockUnaligned(dst + i, 0, endElements);
    }
```

Results:
```
|                        Method |         array | arraySize |      Mean |
|------------------------------ |-------------- |---------- |----------:|
|                     Clear8192 | System.Byte[] |      8192 |  72.56 ns |
|            Clear8192Unaligned | System.Byte[] |      8192 | 102.74 ns |
|     Clear8192Unaligned_JanFix | System.Byte[] |      8192 |  74.92 ns |
|         Clear819Unaligned_Avx | System.Byte[] |      8192 |  56.45 ns |
|                 Clear8192_Avx | System.Byte[] |      8192 |  56.13 ns |
```
I understand it should be tested on different machines, on different test data, including some real world apps so I am leaving it here in case if anybody is interested.

**Pull request merged**